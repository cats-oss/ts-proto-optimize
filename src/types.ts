export type NamespaceReplacement = {
  [key: string]: string | undefined;
};
