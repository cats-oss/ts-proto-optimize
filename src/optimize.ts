import ts from 'typescript';
import { pascalCase } from 'change-case';
import { NamespaceReplacement } from './types';

const nsReplaceIfNeeded = (replacement: NamespaceReplacement, ns: string) => {
  if (replacement.has(ns)) {
    return replacement.get(ns)!;
  }

  return ns;
};

const concatPropertyEnumName = (
  replacement: NamespaceReplacement,
  node: ts.QualifiedName,
): string => {
  const name = nsReplaceIfNeeded(
    replacement,
    node.right.escapedText.toString(),
  );

  if (ts.isQualifiedName(node.left)) {
    return `${concatPropertyEnumName(replacement, node.left)}_${name}`;
  }

  return `${
    ts.isIdentifier(node.left)
      ? `${nsReplaceIfNeeded(replacement, node.left.escapedText.toString())}_`
      : ''
  }${name}`;
};

const updateTypeNodeIfNeeded = (
  replacement: NamespaceReplacement,
  node: ts.TypeNode,
): ts.TypeNode => {
  if (
    ts.isTypeReferenceNode(node) &&
    ts.isQualifiedName(node.typeName) &&
    !/^I[A-Z]/.test(node.typeName.right.escapedText.toString())
  ) {
    return ts.factory.updateTypeReferenceNode(
      node,
      ts.factory.createIdentifier(
        pascalCase(concatPropertyEnumName(replacement, node.typeName)),
      ),
      node.typeArguments,
    );
  }

  return node;
};

const transformer =
  (nsReplacement: NamespaceReplacement): ts.TransformerFactory<ts.Node> =>
  (context) =>
  (source) => {
    const classList: Set<string> = new Set();
    let enumList = ts.factory.createNodeArray<ts.Statement>();

    /**
     * Extract enum
     */
    const extractEnum = (parentNamespace: string) => (node: ts.Node) => {
      const parent = nsReplacement.has(parentNamespace)
        ? nsReplacement.get(parentNamespace)!
        : parentNamespace;

      const ns = parent !== '' ? `${parent}_` : '';

      if (ts.isEnumDeclaration(node)) {
        enumList = ts.factory.createNodeArray([
          ...enumList,
          ts.factory.createEnumDeclaration(
            node.decorators,
            [ts.factory.createModifier(ts.SyntaxKind.ExportKeyword)],
            ts.factory.createIdentifier(pascalCase(`${ns}${node.name.text}`)),
            node.members,
          ),
        ]);
      }

      if (ts.isModuleDeclaration(node) && node.body != null) {
        ts.forEachChild(node.body, extractEnum(`${ns}${node.name.text}`));
      }
    };

    /**
     * Optimize
     */
    const optimizeVisitor: ts.Visitor = (node) => {
      node = ts.visitEachChild(node, optimizeVisitor, context);

      // Remove `protobufjs`
      if (
        ts.isImportDeclaration(node) &&
        ts.isStringLiteral(node.moduleSpecifier) &&
        (<any>node.moduleSpecifier).text === 'protobufjs'
      ) {
        return undefined;
      }

      // Remove `class` definition
      if (ts.isClassDeclaration(node)) {
        classList.add(node.name!.escapedText.toString());
        return undefined;
      }

      // Remove `enum` definition
      if (ts.isEnumDeclaration(node)) {
        return undefined;
      }

      // Format union types
      //   - Remove `null`, `Long`
      //   - Concat `enum` name
      if (ts.isPropertySignature(node) && node.type != null) {
        let unionTypes: ts.NodeArray<ts.TypeNode> | null = null;

        if (ts.isUnionTypeNode(node.type)) {
          unionTypes = node.type.types;
        } else if (
          ts.isParenthesizedTypeNode(node.type) &&
          node.type.type != null &&
          ts.isUnionTypeNode(node.type.type)
        ) {
          unionTypes = node.type.type.types;
        }

        if (unionTypes != null) {
          const unionTypeFilter = (n: ts.Node) => {
            if (
              ts.isLiteralTypeNode(n) &&
              n.literal.kind === ts.SyntaxKind.NullKeyword
            ) {
              return false;
            }

            if (
              ts.isTypeReferenceNode(n) &&
              (<any>n.typeName).escapedText === 'Long'
            ) {
              return false;
            }

            return true;
          };

          const types = unionTypes.filter(unionTypeFilter).map((t) => {
            if (ts.isArrayTypeNode(t)) {
              let elementType = t.elementType;

              if (
                ts.isParenthesizedTypeNode(t.elementType) &&
                ts.isUnionTypeNode(t.elementType.type)
              ) {
                elementType = ts.factory.updateParenthesizedType(
                  t.elementType,
                  ts.factory.updateUnionTypeNode(
                    t.elementType.type,
                    ts.factory.createNodeArray(
                      t.elementType.type.types.filter(unionTypeFilter),
                    ),
                  ),
                );
              }

              return ts.factory.updateArrayTypeNode(
                t,
                updateTypeNodeIfNeeded(nsReplacement, elementType),
              );
            }

            if (ts.isTypeReferenceNode(t)) {
              return updateTypeNodeIfNeeded(nsReplacement, t);
            }

            return t;
          });

          return ts.factory.updatePropertySignature(
            node,
            node.modifiers,
            node.name,
            undefined,
            ts.factory.createParenthesizedType(
              ts.factory.createUnionTypeNode(types),
            ),
          );
        }
      }

      // Add `export` to `namespace`
      if (ts.isModuleDeclaration(node)) {
        return ts.factory.updateModuleDeclaration(
          node,
          node.decorators,
          [ts.factory.createModifier(ts.SyntaxKind.ExportKeyword)],
          node.name,
          node.body,
        );
      }

      // Add `export` to `interface`
      if (ts.isInterfaceDeclaration(node)) {
        return ts.factory.updateInterfaceDeclaration(
          node,
          node.decorators,
          [ts.factory.createModifier(ts.SyntaxKind.ExportKeyword)],
          node.name,
          node.typeParameters,
          node.heritageClauses,
          node.members,
        );
      }

      // Remove type alias (only function type)
      if (ts.isTypeAliasDeclaration(node) && ts.isFunctionTypeNode(node.type)) {
        return undefined;
      }

      return node;
    };

    /**
     * Add declare
     */
    const declareVisitor: ts.Visitor = (node) =>
      ts.visitEachChild(
        node,
        (child) => {
          if (ts.isModuleDeclaration(child)) {
            return ts.factory.updateModuleDeclaration(
              child,
              child.decorators,
              [
                ...child.modifiers!,
                ts.factory.createModifier(ts.SyntaxKind.DeclareKeyword),
              ],
              child.name,
              child.body,
            );
          }
          return child;
        },
        context,
      );

    // Process SourceFile
    source.forEachChild(extractEnum(''));
    source = ts.visitNode(source, optimizeVisitor);
    source = ts.visitNode(source, declareVisitor);

    // Append formatted `enum` list to node statements
    source = ts.factory.updateSourceFile(source as ts.SourceFile, [
      ...(source as ts.SourceFile).statements,
      ...enumList,
    ]);

    return source;
  };

export const optimize = (
  input: string,
  nsReplacement: NamespaceReplacement,
) => {
  const source = ts.createSourceFile('', input, ts.ScriptTarget.Latest, true);
  const { transformed } = ts.transform(source, [transformer(nsReplacement)]);
  return ts.createPrinter().printFile(transformed[0] as any);
};
