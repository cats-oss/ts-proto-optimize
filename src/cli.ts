import path from 'path';
import fs from 'fs';
import meow from 'meow';
import chalk from 'chalk';
import mkdirp from 'mkdirp';
import { optimize } from './optimize';
import { NamespaceReplacement } from './types';

const Status = {
  OK: 0,
  ERROR: 1,
} as const;

const cli = meow(
  `
${chalk.yellow('USAGE:')}
  $ ts-proto-optimize [options] path

${chalk.yellow('OPTIONS:')}
  --output, -o   Path to output file
  --ns           Namespace replacement config (e.g. "before:after")
  --help, -h     Show this help
  --version, -v  Show current version

${chalk.yellow('EXAMPLES:')}
  $ ts-proto-optimize path/to/proto.d.ts
  $ ts-proto-optimize path/to/proto.d.ts --output dist/to/proto.ts
  $ ts-proto-optimize path/to/proto.d.ts --ns root:fuga
  $ ts-proto-optimize path/to/proto.d.ts --ns root:fuga --ns hoge:piyo
  $ ts-proto-optimize path/to/proto.d.ts --ns root:
`,
  {
    flags: {
      output: {
        type: 'string',
        alias: 'o',
      },
      ns: {
        type: 'string',
      },
      help: {
        alias: 'h',
      },
      version: {
        alias: 'v',
      },
    },
  },
);

const fileExists = (input: string) => {
  try {
    return fs.statSync(input).isFile() === true;
  } catch (e) {
    return false;
  }
};

const parseNamespaces = (
  input: string | string[] | undefined,
): NamespaceReplacement => {
  const map = new Map();

  if (input == null) {
    return map;
  }

  if (Array.isArray(input)) {
    input
      .map((v) => parseNamespaces(v))
      .forEach((m) => {
        m.forEach((v, k) => {
          map.set(k, v);
        });
      });

    return map;
  }

  const [key, value] = input.split(':');

  map.set(key, value);

  return map;
};

(async () => {
  try {
    const cwd = process.cwd();

    const filepath = path.resolve(cwd, cli.input[0]);
    if (!fileExists(filepath)) {
      throw new Error(`proto file does not exist ("${filepath}")`);
    }

    const output =
      cli.flags.output != null ? path.join(cwd, cli.flags.output) : null;

    const input = fs.readFileSync(filepath, { encoding: 'utf8' });
    const result = optimize(input, parseNamespaces(cli.flags.ns));

    if (output == null) {
      console.log(result);
    } else {
      mkdirp.sync(path.dirname(output));
      fs.writeFileSync(output, result);
    }

    process.exit(Status.OK);
  } catch (e) {
    console.error(
      `${chalk.bgRed.bold.white(' ERROR ')} ${chalk.red(e.message)}`,
    );

    process.exit(Status.ERROR);
  }
})();
