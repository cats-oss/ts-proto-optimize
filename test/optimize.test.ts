import path from 'path';
import fs from 'fs';
import { optimize } from '../src/optimize';

const input = fs.readFileSync(path.resolve(__dirname, '../fixtures/proto.d.ts'), 'utf-8');

test('optimize - without ns', () => {
  expect(optimize(input, {})).toMatchSnapshot();
});

test('optimize - with ns', () => {
  expect(optimize(input, { root: '', api: 'example' })).toMatchSnapshot();
});
