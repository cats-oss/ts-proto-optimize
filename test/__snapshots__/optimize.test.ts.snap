// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`optimize - with ns 1`] = `
"/** Namespace root. */
export declare namespace root {
    /** Namespace api. */
    export namespace api {
        /** Properties of a Foo */
        export interface IFoo {
            /** Foo str */
            str: (string);
            /** Foo bool */
            bool: (boolean);
            /** Foo number */
            num: (number);
            /** Foo metadata */
            metadata: ({
                [k: string]: string;
            });
        }
        export namespace AdminService {
            /** Properties of a Enumerate */
            export interface IEnumerate {
                /** Enumerate id */
                id: (string);
                /** Enumrate field */
                field: (ExampleEnum);
                /** Enumrate arr */
                arr: (ExampleEnum[]);
            }
        }
        /** Properties of a Enumerate */
        export interface IEnumerate {
            /** Enumerate id */
            id: (string);
            /** Enumrate field */
            field: (ExampleEnum);
            /** Enumrate arr */
            arr: (ExampleEnum[]);
        }
        export namespace Nest {
            /** Properties Data */
            export interface IData {
                name: (string);
                value: (number);
                field: (root.api.Nest.IData);
                enumratee: (ExampleNestEnum);
            }
        }
    }
}
/** Namespace depth1. */
export declare namespace depth1 {
    export namespace depth2 {
        export namespace depth3 {
            export interface IFoo {
                /** Foo key */
                key: (string);
            }
        }
    }
}
export enum ExampleAdminServiceEnum {
    KEY1 = 0,
    KEY2 = 1,
    KEY3 = 2
}
export enum ExampleEnum {
    KEY1_SNAKE_CASE = 0,
    KEY2_SNAKE_CASE = 1,
    KEY3_SNAKE_CASE = 2
}
export enum ExampleNestEnum {
    KEY1 = 0,
    KEY2 = 1,
    KEY3 = 2
}
"
`;

exports[`optimize - without ns 1`] = `
"/** Namespace root. */
export declare namespace root {
    /** Namespace api. */
    export namespace api {
        /** Properties of a Foo */
        export interface IFoo {
            /** Foo str */
            str: (string);
            /** Foo bool */
            bool: (boolean);
            /** Foo number */
            num: (number);
            /** Foo metadata */
            metadata: ({
                [k: string]: string;
            });
        }
        export namespace AdminService {
            /** Properties of a Enumerate */
            export interface IEnumerate {
                /** Enumerate id */
                id: (string);
                /** Enumrate field */
                field: (RootApiEnum);
                /** Enumrate arr */
                arr: (RootApiEnum[]);
            }
        }
        /** Properties of a Enumerate */
        export interface IEnumerate {
            /** Enumerate id */
            id: (string);
            /** Enumrate field */
            field: (RootApiEnum);
            /** Enumrate arr */
            arr: (RootApiEnum[]);
        }
        export namespace Nest {
            /** Properties Data */
            export interface IData {
                name: (string);
                value: (number);
                field: (root.api.Nest.IData);
                enumratee: (RootApiNestEnum);
            }
        }
    }
}
/** Namespace depth1. */
export declare namespace depth1 {
    export namespace depth2 {
        export namespace depth3 {
            export interface IFoo {
                /** Foo key */
                key: (string);
            }
        }
    }
}
export enum RootApiAdminServiceEnum {
    KEY1 = 0,
    KEY2 = 1,
    KEY3 = 2
}
export enum RootApiEnum {
    KEY1_SNAKE_CASE = 0,
    KEY2_SNAKE_CASE = 1,
    KEY3_SNAKE_CASE = 2
}
export enum RootApiNestEnum {
    KEY1 = 0,
    KEY2 = 1,
    KEY3 = 2
}
"
`;
